### PR TITLE
[JENKINS-72833] Do not attempt to self-`exec` on systems without `libc`

### PR DESCRIPTION
**Pull request by basil**
See [JENKINS-72833](https://issues.jenkins.io/browse/JENKINS-72833). We were attempting to use `UnixLifecycle` on a system not supported by JNA.

While I was here I also did some other cleanup of this class.

### Testing done

Tested `/safeRestart` on Linux both inside and outside `systemd`.

### Proposed changelog entries

- Do not attempt to self-restart on operating systems where this is not supported.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/lifecycle/Lifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/Lifecycle.java:123]: Dead code, since this constructor could never in fact throw `IOException`, but I thought it best to keep this fallback since any unanticipated errors here should not be fatal.

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java:56]: This can never be null, so annotate it as such.

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java:56]: This can never throw `IOException`, so stop declaring it.

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java:62]: This was based on the old implementation, which used Akuma. That was ripped out a long time ago, and these exceptions shouldn't be thrown in the current implementation. But just to be on the safe side, we're now catching `Throwable` in the caller, so this shouldn't result in a regression in any case.

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java:92]: The fix for JENKINS-72833: do not attempt to use `UnixLifecycle` if JNA can't make calls to `libc`.

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (basil) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java:101]: Dead code, since `args` can never be null.

File: core/src/main/java/jenkins/util/JavaVMArguments.java
Comment (basil) [inline core/src/main/java/jenkins/util/JavaVMArguments.java:34]: Unnecessary use of Guava.

**Review by MarkEWaite — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Review by StefanSpieker — APPROVED** [review]

**Pull request merged**
